### PR TITLE
Enable dnstap plugin to insert other plugin's specific data into extra field of tap.Dnstap message

### PR DESCRIPTION
**Pull request by huynhmb359**
### 1. What does this pull request do
This PR enables dnstap plugin to insert other plugin's specific data into extra field of tap.Dnstap message.    

- Dnstap allocates a map where other plugins can add their extra data.   Key is the plugin name and value is the DnstapExtra.    DnstapExtra is a map where extra data per message type are stored.  Key is the message type (i.e.  CQ,  CR  ...etc) .  Value is the wire data in byte array.  

- We define a new structure where the extra data for various plugins can be written into the tap.Dnstap's extra using protobuf.   The structure called Extra.  It's map  with key is the name of the plugin and value is the extra data. 

- When sending out the dnstap message,  dnstap plugin will use the data in the map in the context to construct Extra data and then sends in the tap.DnsTap. 

 2. Which issues (if any) are related?
There is no issue opened for this.

### 3. Which documentation changes (if any) need to be made?
No document change

I'll add UTs after the approach accepted.

### CODE REVIEW

**Comment by miekg** [comment]
/cc @varyoo 

I think this can work. Making a new map on every connection is somewhat expensive though - what amount of data can actually be put here? I only see a `[]byte`.

**Comment by huynhmb359** [comment]
For now it's only []byte.  Middleware can put its own serialize data here.

**Comment by 2buisson** [comment]
If a middleware wants to send a custom dnstap client response maybe it should do it completely and signal the dnstap middleware not to do it before the response is written.

**Comment by johnbelamaric** [comment]
@varyoo in that case I guess we can get rid of the map creation. on the other hand, only one middleware can add that data, rather than accumulating stuff from all the middleware.

As of today, that works for our use case - we only need to do it with one middleware.

@huynhmb359 can we create the map only if we need it - that is, in the policy MW? Though, then we take a perf hit which we won't in @varyoo's solution. Hmm..

I guess given that we don't *need* multiple middleware to do it today, and there is a performance cost to doing that, we are better off taking @varyoo's suggestion. I think that means sticking a boolean in the context, and having the dnstap plugin skip the CR in that case. Make sense?

**Comment by huynhmb359** [comment]
OK, I will fix according to both @varyoo  and @johnbelamaric  suggestion.

**Comment by miekg** [comment]
you can also just append to a `[][]byte` in the context?

**Comment by huynhmb359** [comment]
Not sure why my recent commit :  https://github.com/huynhmb359/coredns/commit/8b33d0a26b591d7e7637113fac705af7beae361c  
does not go into this PR?    

Any suggestion?   Should I close this PR and recreate?

File: plugin/dnstap/taprw/writer.go [hunk 1]
```diff
@@ -19,12 +26,15 @@ type Tapper interface {
 
 // ResponseWriter captures the client response and logs the query to dnstap.
 // Single request use.
+// DnsTapExtras is map of DnstapExtra.  Key is the name of the middlware that
+// saves the extra data.  Value is a DnstapExtra.
```
Comment (johnbelamaric) [inline plugin/dnstap/taprw/writer.go]: Fix the comment.

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
Nit: Just fix the comment then this LGTM.

File: plugin/dnstap/taprw/writer.go (hunk 1 above)
Comment (huynhmb359) [reply plugin/dnstap/taprw/writer.go]: It's fixed.  Thnx.

**Comment by johnbelamaric** [comment]
I think the test failed due to this merging in upstream miekg/dns: https://github.com/miekg/dns/pull/513

**Pull request merged**